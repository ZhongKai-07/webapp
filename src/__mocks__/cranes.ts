import { FeatureCollection, Point } from 'geojson';

export const cranes: FeatureCollection<Point> = {
  type: 'FeatureCollection',
  features: [
    {
      type: 'Feature',
      properties: {
        craneId: 1,
        height: 50,
        radius: 10,
      },
      geometry: {
        type: 'Point',
        coordinates: [0, 0],
      },
    },
    {
      type: 'Feature',
      properties: {
        craneId: 2,
        height: 40,
        radius: 8,
      },
      geometry: {
        type: 'Point',
        coordinates: [1, 1],
      },
    },
  ],
};
