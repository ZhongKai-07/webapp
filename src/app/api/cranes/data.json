{
  "type": "FeatureCollection",
  "features": [
    {
      "type": "Feature",
      "properties": {
        "craneId": "1",
        "height": 187,
        "radius": 122,
        "timestamp": 1705232121,
        "image": "https://images.pexels.com/photos/7245353/pexels-photo-7245353.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2&index=0"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.11433123536651221, 51.497263819259494]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "2",
        "height": 275,
        "radius": 109,
        "timestamp": 1705232121,
        "image": "https://t3.ftcdn.net/jpg/00/71/80/08/360_F_71800864_SvUJAXhYoHVxInt52qFLaf6oxuuc32gL.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.1149200627442593, 51.4970587122]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "3",
        "height": 53,
        "radius": 91,
        "timestamp": 1705232121,
        "image": "https://c8.alamy.com/comp/EXMT2E/cranes-in-the-city-of-london-EXMT2E.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.09874059769622952, 51.51428631981964]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "4",
        "height": 234,
        "radius": 104,
        "timestamp": 1705232121,
        "image": "https://www.shutterstock.com/shutterstock/photos/2353031057/display_1500/stock-photo-london-uk-august-st-pauls-cathedral-surrounded-by-construction-cranes-2353031057.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10384662070591874, 51.506350672575515]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "5",
        "height": 77,
        "radius": 35,
        "timestamp": 1705232121,
        "image": "https://media.istockphoto.com/id/1167905876/photo/london-citys-financial-district-skyline-with-all-iconic-buildings-and-construction-cranes.jpg?s=612x612&w=0&k=20&c=GXwFamCRh7DlNPm-qzus84i9JZsJlbsP-VHqYuUmVsA="
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.1218620467606042, 51.504695183961275]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "6",
        "height": 61,
        "radius": 75,
        "timestamp": 1705280394,
        "image": "https://c8.alamy.com/comp/EXMT2E/cranes-in-the-city-of-london-EXMT2E.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10701641144000305, 51.50180701131141]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "7",
        "height": 53,
        "radius": 127,
        "timestamp": 1705280394,
        "image": "https://c8.alamy.com/comp/EXMT2E/cranes-in-the-city-of-london-EXMT2E.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.09425554153211935, 51.50517805771138]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "8",
        "height": 221,
        "radius": 79,
        "timestamp": 1705280394,
        "image": "https://c8.alamy.com/comp/EXMT2E/cranes-in-the-city-of-london-EXMT2E.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10607721553295997, 51.5005941236613]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "9",
        "height": 160,
        "radius": 47,
        "timestamp": 1705280394,
        "image": "https://c8.alamy.com/comp/EXMT2E/cranes-in-the-city-of-london-EXMT2E.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.11766537868031536, 51.509112593520655]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "10",
        "height": 64,
        "radius": 84,
        "timestamp": 1705280394,
        "image": "https://c8.alamy.com/comp/EXMT2E/cranes-in-the-city-of-london-EXMT2E.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.11117899308378665, 51.50305965491295]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "11",
        "height": 243,
        "radius": 120,
        "timestamp": 1705354435,
        "image": "https://images.pexels.com/photos/7245353/pexels-photo-7245353.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2&index=3"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10536901340133668, 51.496558078576946]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "12",
        "height": 100,
        "radius": 90,
        "timestamp": 1705354435,
        "image": "https://images.pexels.com/photos/7245353/pexels-photo-7245353.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2&index=3"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10702058353176155, 51.51333735639661]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "13",
        "height": 158,
        "radius": 61,
        "timestamp": 1705354435,
        "image": "https://images.pexels.com/photos/7245353/pexels-photo-7245353.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2&index=3"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.1013700192375014, 51.49987945696387]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "14",
        "height": 177,
        "radius": 118,
        "timestamp": 1705354435,
        "image": "https://images.pexels.com/photos/7245353/pexels-photo-7245353.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2&index=3"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10161997647498605, 51.51054792156448]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "15",
        "height": 221,
        "radius": 57,
        "timestamp": 1705354435,
        "image": "https://images.pexels.com/photos/7245353/pexels-photo-7245353.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2&index=3"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.12475318939258816, 51.50132321254501]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "16",
        "height": 123,
        "radius": 35,
        "timestamp": 1705307333,
        "image": "https://www.shutterstock.com/shutterstock/photos/2353031057/display_1500/stock-photo-london-uk-august-st-pauls-cathedral-surrounded-by-construction-cranes-2353031057.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10984847503983988, 51.512904831395254]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "17",
        "height": 145,
        "radius": 113,
        "timestamp": 1705307333,
        "image": "https://www.shutterstock.com/shutterstock/photos/2353031057/display_1500/stock-photo-london-uk-august-st-pauls-cathedral-surrounded-by-construction-cranes-2353031057.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.10139748495285902, 51.497345725257]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "18",
        "height": 125,
        "radius": 96,
        "timestamp": 1705307333,
        "image": "https://www.shutterstock.com/shutterstock/photos/2353031057/display_1500/stock-photo-london-uk-august-st-pauls-cathedral-surrounded-by-construction-cranes-2353031057.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.1200828901143363, 51.50100606850154]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "19",
        "height": 129,
        "radius": 137,
        "timestamp": 1705307333,
        "image": "https://www.shutterstock.com/shutterstock/photos/2353031057/display_1500/stock-photo-london-uk-august-st-pauls-cathedral-surrounded-by-construction-cranes-2353031057.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.12518775866199497, 51.50449288520402]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "20",
        "height": 279,
        "radius": 90,
        "timestamp": 1705307333,
        "image": "https://www.shutterstock.com/shutterstock/photos/2353031057/display_1500/stock-photo-london-uk-august-st-pauls-cathedral-surrounded-by-construction-cranes-2353031057.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.09943138225742051, 51.49820852900131]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "21",
        "height": 147,
        "radius": 74,
        "timestamp": 1705383194,
        "image": "https://t3.ftcdn.net/jpg/00/71/80/08/360_F_71800864_SvUJAXhYoHVxInt52qFLaf6oxuuc32gL.jpg"
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.1174855545800235, 51.5008315750743]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "22",
        "height": 144,
        "radius": 67,
        "timestamp": 1705383194,
        "image": "https://media.istockphoto.com/id/1167905876/photo/london-citys-financial-district-skyline-with-all-iconic-buildings-and-construction-cranes.jpg?s=612x612&w=0&k=20&c=GXwFamCRh7DlNPm-qzus84i9JZsJlbsP-VHqYuUmVsA="
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.12003226436123904, 51.510042959039296]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "23",
        "height": 243,
        "radius": 90,
        "timestamp": 1705383194,
        "image": "https://media.istockphoto.com/id/1167905876/photo/london-citys-financial-district-skyline-with-all-iconic-buildings-and-construction-cranes.jpg?s=612x612&w=0&k=20&c=GXwFamCRh7DlNPm-qzus84i9JZsJlbsP-VHqYuUmVsA="
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.11981069161203699, 51.508453467509696]
      }
    },
    {
      "type": "Feature",
      "properties": {
        "craneId": "24",
        "height": 124,
        "radius": 102,
        "timestamp": 1705383194,
        "image": "https://media.istockphoto.com/id/1167905876/photo/london-citys-financial-district-skyline-with-all-iconic-buildings-and-construction-cranes.jpg?s=612x612&w=0&k=20&c=GXwFamCRh7DlNPm-qzus84i9JZsJlbsP-VHqYuUmVsA="
      },
      "geometry": {
        "type": "Point",
        "coordinates": [-0.09281834099061975, 51.50057585662843]
      }
    }
  ]
}