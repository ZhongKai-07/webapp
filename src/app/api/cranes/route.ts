import { FeatureCollection, Point } from 'geojson';
import { NextResponse } from 'next/server';

export async function GET() {
  try {
    const fileData = await import('./data.json');
    const dropZoneMockData: FeatureCollection<Point> =
      fileData as FeatureCollection<Point>;
    return NextResponse.json(dropZoneMockData);
  } catch (error) {
    return NextResponse.json({});
  }
}
