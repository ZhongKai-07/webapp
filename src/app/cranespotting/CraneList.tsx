/* eslint-disable @next/next/no-img-element */
import { FeatureCollection, Point } from 'geojson';
import React from 'react';

interface CraneListProps {
  cranes?: FeatureCollection<Point>;
}

const CraneList: React.FC<CraneListProps> = ({ cranes }) => {
  const sortedCranes = cranes
    ? {
        type: cranes.type,
        features: cranes.features.sort((a, b) => {
          const idA: number = a.properties?.craneId || 0;
          const idB: number = b.properties?.craneId || 0;
          return idA - idB;
        }),
      }
    : null;

  return cranes && cranes?.features?.length > 0 ? (
    <div className='overflow-auto p-2 w-100'>
      <ul className='m-0 list-none p-0 grid grid-cols-3 grid-rows-15 gap-2'>
        {sortedCranes &&
          sortedCranes.features.map(
            (feature) =>
              feature && (
                <li
                  key={feature?.properties?.craneId}
                  className='mb-2 rounded bg-white p-2 shadow-box-grey-sm'
                >
                  <div className='font-bold'>
                    Crane {feature?.properties?.craneId}
                  </div>
                  <div>Height: {feature.properties?.height}m</div>
                  <div>Radius: {feature.properties?.radius}m</div>
                </li>
              )
          )}
      </ul>
    </div>
  ) : (
    <div className='flex w-full h-full overflow-hidden justify-between'>
      <img
        className='ml-[-8em] w-[70%] mt-[-4em]'
        width={0}
        height={0}
        src='/images/missing_crane.png'
        alt='image of a crane'
      />
      <p className='w-[48%]  pr-[1em] text-5xl'>
        Select some timestamps to view crane data and images.
      </p>
    </div>
  );
};

export default CraneList;
