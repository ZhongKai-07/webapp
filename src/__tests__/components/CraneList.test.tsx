import { render } from '@testing-library/react';

import { cranes } from '@/__mocks__/cranes';
import CraneList from '@/app/cranespotting/CraneList';

test('renders the list of cranes correctly', () => {
  const { asFragment } = render(<CraneList cranes={cranes} />);
  expect(asFragment()).toMatchSnapshot();
});
