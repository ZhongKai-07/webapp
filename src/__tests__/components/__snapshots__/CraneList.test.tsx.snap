// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the list of cranes correctly 1`] = `
<DocumentFragment>
  <div
    class="overflow-auto p-2 w-100"
  >
    <ul
      class="m-0 list-none p-0 grid grid-cols-3 grid-rows-15 gap-2"
    >
      <li
        class="mb-2 rounded bg-white p-2 shadow-box-grey-sm"
      >
        <div
          class="font-bold"
        >
          Crane 1
        </div>
        <div>
          Height: 50m
        </div>
        <div>
          Radius: 10m
        </div>
      </li>
      <li
        class="mb-2 rounded bg-white p-2 shadow-box-grey-sm"
      >
        <div
          class="font-bold"
        >
          Crane 2
        </div>
        <div>
          Height: 40m
        </div>
        <div>
          Radius: 8m
        </div>
      </li>
    </ul>
  </div>
</DocumentFragment>
`;
